File: Week 1/Problem Set 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int checksum(long n);
int nodigits(long n);
void identify(long n);

int main(void)
{
    long n;
    n = get_long("Number: ");
    if (checksum(n) == 0)
    {
        printf("INVALID\n");
    }
    else
    {
        identify(n);
    }
}

int nodigits(long n)
{
    long temp = n;
    int i = 0;
    while (temp)
    {
        temp /= 10;
        i++;
    }

    return i;
}

void identify(long n)
{
    if (nodigits(n) == 15)
    {
        if (n / 10000000000000 == 34 || n / 10000000000000 == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (nodigits(n) == 16)
    {
        long tmp = n / 100000000000000;
        if (n / 1000000000000000 == 4)
        {
            printf("VISA\n");
        }
        else if (tmp == 51 || tmp == 52 || tmp == 53 || tmp == 54 || tmp == 55)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (nodigits(n) == 13)
    {
        if (n / 1000000000000 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int checksum(long n)
{
    long temp = n;
    int sum = 0;
    int i = 0;
    while (temp)
    {
        if (i == 0)
        {
            sum += temp % 10;
            i = 1;
        }
        else if (i == 1)
        {
            if ((temp % 10) * 2 > 9)
            {
                sum = sum + ((temp % 10) * 2) / 10 + ((temp % 10) * 2) % 10;
            }
            else
            {
                sum = sum + (temp % 10) * 2;
            }
            i = 0;
        }
        temp /= 10;
    }
    if (sum % 10 == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

File: Week 6/Problem Set 6/sentimental-credit/credit.py
def main():
    while True:
        try:
            card = input("Number: ")
            int(card)
        except ValueError:
            continue
        else:
            break
    if checksum(card) != 0:
        print("INVALID")
    else:
        print(identify(card))


def checksum(str):
    list = str[-2::-2]
    sum = 0
    for i in list:
        number = int(i) * 2
        if number > 9:
            sum += int(number / 10) + int(number % 10)
        else:
            sum += number

    list = str[-1::-2]
    for i in list:
        sum += int(i)

    return sum % 10


def identify(n):
    if len(n) == 15:
        if n[:2] == "34" or n[:2] == "37":
            return "AMEX"
        else:
            return "INVALID"
    elif len(n) == 16:
        if n[:1] == "4":
            return "VISA"
        elif n[0] == "5" and n[1] in ["1", "2", "3", "4", "5"]:
            return "MASTERCARD"
        else:
            return "INVALID"
    elif len(n) == 13:
        if n[:1] == "4":
            return "VISA"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
